src/com/amazon/corretto/crypto/provider/ConstantTime.java

```diff
@@ -51,4 +59,53 @@ static final int select(int selector, int a, int b) {
 
     return b ^ (combined & mask);
   }
+
+  /**
+   * All the bytes in the trusted buffer between trustedStart and trustedStart + trustedLen are
+   * examined if otherLen != 0. Please make sure the trusted array is the one that we trust.
+   *
+   * @param trusted
+   * @param trustedStart
+   * @param trustedLen
+   * @param other
+   * @param otherStart
+   * @param otherLen
+   * @return true iff all the bytes in the specified ranges are equal
+   */
+  static boolean equals(
+      final byte[] trusted,
+      final int trustedStart,
+      final int trustedLen,
+      final byte[] other,
+      final int otherStart,
+      final int otherLen) {
+    if (otherLen == 0) {
+      return trustedLen == 0;
+    }
+
+    int result = trustedLen - otherLen;
+
+    for (int i = 0; i < trustedLen; i++) {
+      final int trustedIndex = trustedStart + i;
+      final int otherIndex = otherStart + gtPositive(otherLen, i) * i;
+      result |= trusted[trustedIndex] ^ other[otherIndex];
+    }
+
+    return result == 0;
+  }
+
+  static boolean equals(
+      final byte[] trusted, final int trustedStart, final int trustedLen, final byte[] other) {
+    return equals(trusted, trustedStart, trustedLen, other, 0, other.length);
+  }
+
+  static boolean equals(final byte[] trusted, final byte[] other) {
+    if (trusted == other) {
+      return true;
+    }
+    if (trusted == null || other == null) {
```

Comment:
we could also return false here if `trusted.length != other.length`, right?

Reply:
It could. I moved it to the "main" one.